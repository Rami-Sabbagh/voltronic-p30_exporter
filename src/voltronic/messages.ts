import { crc16Xmodem } from './crc16-xmodem';
import { ChecksumMismatch, InvalidMessage } from './errors';

/**
 * Packs the message into the transmission ready format of voltronics.
 * 
 * Calculates the checksum of the message and appends it to the end of the message
 * alongside a carriage return as specified by the protocol.
 */
export function packMessage(message: string): Buffer {
    const result = Buffer.alloc(message.length + 3);
    result.write(message, 'ascii');

    const checksum = crc16Xmodem(result.subarray(0, -3));
    result.writeUInt16BE(checksum, message.length);
    result.write('\r', result.length - 1, 'ascii');

    return result;
}

/**
 * Unpack the message from the transmitted format.
 * 
 * Validates the checksum and strips it alongside the carriage return from the
 * message as specified by the protocol.
 * 
 * Also removes the first character from the message if it was `(`.
 */
export function unpackMessage(data: Buffer): string;
export function unpackMessage(data: Buffer, raw: true): Buffer;
export function unpackMessage(data: Buffer, raw?: true): string | Buffer {
    if (data.at(-1) !== 0x0D) throw new InvalidMessage('Missing carriage return.');
    if (data.length < 4) throw new InvalidMessage('Data is too short to contain a message.');

    const expected = data.readUInt16BE(data.length - 3);
    const actual = crc16Xmodem(data.subarray(0, data.length - 3));

    if (expected !== actual)
        throw new ChecksumMismatch(expected, actual);
    
    if (raw) return data.subarray(0, data.length - 3);
    return data.toString('ascii', 0, data.length - 3);
}
