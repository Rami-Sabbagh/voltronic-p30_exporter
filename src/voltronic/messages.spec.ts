import { packMessage, unpackMessage } from './messages';

test.each([
    ['QPI', '51 50 49 be ac 0d'],
    ['VOLTRONICS', '56 4f 4c 54 52 4f 4e 49 43 53 6c 3e 0d'],
    ['#$!@-,;:[]/', '23 24 21 40 2d 2c 3b 3a 5b 5d 2f e1 a2 0d']
])('packMessage("%s") is [%s]', (message: string, hex: string) => {
    const expected = Buffer.from(hex.replaceAll(' ', ''), 'hex');
    const actual = packMessage(message);

    expect(actual).toStrictEqual(expected);
});

test.each([
    ['51 50 49 BE AC 0D', 'QPI'],
    ['56 4f 4c 54 52 4f 4e 49 43 53 6c 3e 0d', 'VOLTRONICS'],
    ['23 24 21 40 2d 2c 3b 3a 5b 5d 2f e1 a2 0d', '#$!@-,;:[]/'],
])('unpackMessage([%s]) is "%s"', (hex: string, message: string) => {
    const data = Buffer.from(hex.replaceAll(' ', ''), 'hex');
    const actual = unpackMessage(data);

    expect(actual).toStrictEqual(message);
});

test('Checksum fails with a missing byte', () => {
    const hex = '50 49 BE AC 0D' // QPI but missing the first byte.
    const data = Buffer.from(hex.replaceAll(' ', ''), 'hex');
    
    expect(() => unpackMessage(data)).toThrow(/^Checksum mismatch:/g);
});

test('Checksum fails with a corrupted byte', () => {
    const hex = '51 B9 49 BE AC 0D' // QPI but with P messed up.
    const data = Buffer.from(hex.replaceAll(' ', ''), 'hex');
    
    expect(() => unpackMessage(data)).toThrow(/^Checksum mismatch:/g);
});

test('Missing carriage return is rejected', () => {
    const hex = '51 50 49 BE AC' // QPI missing the carriage return.
    const data = Buffer.from(hex.replaceAll(' ', ''), 'hex');
    
    expect(() => unpackMessage(data)).toThrow('Missing carriage return.');
});

test.each([
    ['0D'],
    ['AC 0D'],
    ['BE AC 0D'],
])('[%s] is rejected for being too short', (hex: string) => {
    const data = Buffer.from(hex.replaceAll(' ', ''), 'hex');
    expect(() => unpackMessage(data)).toThrow('Data is too short to contain a message.');
});

test.each([
    ['Hello World!'],
    ['The Little Quick Brown Fox Jumps Over The Lazy Dog.'],
])('"%s" packs, unpacks and repacks as expected', (message: string) => {
    const packed = packMessage(message);
    expect(packed).toMatchSnapshot();

    const unpacked = unpackMessage(packed);
    expect(unpacked).toStrictEqual(message);

    const repacked = packMessage(unpacked);
    expect(repacked).toStrictEqual(packed);
});

test('Leading "(" is not stripped', () => {
    const packed = packMessage('(ACK');
    const unpacked = unpackMessage(packed);
    expect(unpacked).toStrictEqual('(ACK');
});
