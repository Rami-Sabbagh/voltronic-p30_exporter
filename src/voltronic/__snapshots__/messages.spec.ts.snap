// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"Hello World!" packs, unpacks and repacks as expected 1`] = `
{
  "data": [
    72,
    101,
    108,
    108,
    111,
    32,
    87,
    111,
    114,
    108,
    100,
    33,
    12,
    211,
    13,
  ],
  "type": "Buffer",
}
`;

exports[`"The Little Quick Brown Fox Jumps Over The Lazy Dog." packs, unpacks and repacks as expected 1`] = `
{
  "data": [
    84,
    104,
    101,
    32,
    76,
    105,
    116,
    116,
    108,
    101,
    32,
    81,
    117,
    105,
    99,
    107,
    32,
    66,
    114,
    111,
    119,
    110,
    32,
    70,
    111,
    120,
    32,
    74,
    117,
    109,
    112,
    115,
    32,
    79,
    118,
    101,
    114,
    32,
    84,
    104,
    101,
    32,
    76,
    97,
    122,
    121,
    32,
    68,
    111,
    103,
    46,
    135,
    143,
    13,
  ],
  "type": "Buffer",
}
`;
